File: src/ainfra/contracts.py
"""Load and validate ainfra v1alpha1 contract documents."""

from __future__ import annotations

import json
from ipaddress import IPv4Network, IPv6Network
from pathlib import Path
from typing import Any, cast

import yaml  # type: ignore[import-untyped]
from jsonschema import (  # type: ignore[import-untyped]
    Draft202012Validator,
    FormatChecker,
)

from ainfra.errors import ContractError

API_VERSION = "ainfra.projectious.work/v1alpha1"
FORMAT_CHECKER = FormatChecker()
SCHEMA_FILES = {
    "InfrastructureTemplate": "template-manifest.v1alpha1.json",
    "TemplateInput": "template-input.v1alpha1.json",
    "InfrastructureOutput": "template-output.v1alpha1.json",
}


@FORMAT_CHECKER.checks("ipv4-network", raises=ValueError)  # type: ignore[misc]
def _is_ipv4_network(value: object) -> bool:
    if not isinstance(value, str):
        return False
    IPv4Network(value, strict=True)
    return True


@FORMAT_CHECKER.checks("ipv6-network", raises=ValueError)  # type: ignore[misc]
def _is_ipv6_network(value: object) -> bool:
    if not isinstance(value, str):
        return False
    IPv6Network(value, strict=True)
    return True


def repository_root() -> Path:
    """Return the source checkout root containing the public schemas."""

    return Path(__file__).resolve().parents[2]


def load_document(path: Path) -> dict[str, Any]:
    """Read a JSON or YAML object without resolving any secret reference."""

    try:
        content = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise ContractError(f"cannot read {path}: {exc}") from exc

    try:
        value = (
            json.loads(content)
            if path.suffix.lower() == ".json"
            else yaml.safe_load(content)
        )
    except (json.JSONDecodeError, yaml.YAMLError) as exc:
        raise ContractError(f"cannot parse {path}: {exc}") from exc

    if not isinstance(value, dict):
        raise ContractError(f"{path}: document must be an object")
    return value


def schema_for(kind: str) -> dict[str, Any]:
    """Load the schema associated with a supported document kind."""

    filename = SCHEMA_FILES.get(kind)
    if filename is None:
        supported = ", ".join(sorted(SCHEMA_FILES))
        raise ContractError(f"unsupported kind {kind!r}; expected {supported}")
    path = repository_root() / "schemas" / filename
    return cast(
        dict[str, Any],
        json.loads(path.read_text(encoding="utf-8")),
    )


def validate_document(document: dict[str, Any], *, source: Path) -> None:
    """Validate a document and report the first deterministic error."""

    kind = document.get("kind")
    if not isinstance(kind, str):
        raise ContractError(f"{source}: /kind is required and must be a string")

    validator = Draft202012Validator(
        schema_for(kind),
        format_checker=FORMAT_CHECKER,
    )
    errors = sorted(
        validator.iter_errors(document), key=lambda err: list(err.path)
    )
    if not errors:
        return

    error = errors[0]
    pointer = "/" + "/".join(str(part) for part in error.absolute_path)
    if pointer == "/":
        pointer = "<root>"
    is_output = kind == "InfrastructureOutput"
    raise ContractError(
        f"{source}: {pointer}: {error.message}",
        output=is_output,
    )


def validate_path(path: Path) -> dict[str, Any]:
    """Load and validate one public contract document."""

    document = load_document(path)
    validate_document(document, source=path)
    return document

File: src/contracts.rs
//! Embedded `v1alpha1` contract validation.

use std::fs;
use std::net::{Ipv4Addr, Ipv6Addr};
use std::path::Path;

use jsonschema::Draft;
use serde_json::Value;

use crate::error::AinfraError;

const MANIFEST_SCHEMA: &str = include_str!("../schemas/template-manifest.v1alpha1.json");
const INPUT_SCHEMA: &str = include_str!("../schemas/template-input.v1alpha1.json");
const OUTPUT_SCHEMA: &str = include_str!("../schemas/template-output.v1alpha1.json");

/// Load and validate one JSON or YAML contract document.
///
/// # Errors
///
/// Returns a stable input/output contract error for read, parse, dispatch, or
/// schema validation failures.
pub fn validate_path(path: &Path) -> Result<Value, AinfraError> {
    let content = fs::read_to_string(path).map_err(|error| {
        AinfraError::input_contract(format!("cannot read {}: {error}", path.display()))
    })?;
    let document = if path.extension().is_some_and(|suffix| suffix == "json") {
        serde_json::from_str(&content).map_err(|error| {
            AinfraError::input_contract(format!("cannot parse {}: {error}", path.display()))
        })?
    } else {
        serde_yaml::from_str(&content).map_err(|error| {
            AinfraError::input_contract(format!("cannot parse {}: {error}", path.display()))
        })?
    };
    validate_document(&document, path)?;
    Ok(document)
}

/// Validate an already parsed contract document.
///
/// # Errors
///
/// Returns a stable input/output contract error when the kind or schema is
/// invalid.
pub fn validate_document(document: &Value, source: &Path) -> Result<(), AinfraError> {
    let object = document.as_object().ok_or_else(|| {
        AinfraError::input_contract(format!("{}: document must be an object", source.display()))
    })?;
    let kind = object.get("kind").and_then(Value::as_str).ok_or_else(|| {
        AinfraError::input_contract(format!(
            "{}: /kind is required and must be a string",
            source.display()
        ))
    })?;
    let schema_text = match kind {
        "InfrastructureTemplate" => MANIFEST_SCHEMA,
        "TemplateInput" => INPUT_SCHEMA,
        "InfrastructureOutput" => OUTPUT_SCHEMA,
        other => {
            return Err(AinfraError::input_contract(format!(
                "unsupported kind {other:?}; expected InfrastructureOutput, \
                 InfrastructureTemplate, TemplateInput"
            )));
        }
    };
    let schema: Value = serde_json::from_str(schema_text)
        .map_err(|error| AinfraError::dependency(format!("embedded schema is invalid: {error}")))?;
    let validator = jsonschema::options()
        .with_draft(Draft::Draft202012)
        .should_validate_formats(true)
        .with_format("ipv4-network", is_ipv4_network)
        .with_format("ipv6-network", is_ipv6_network)
        .build(&schema)
        .map_err(|error| AinfraError::dependency(format!("embedded schema is invalid: {error}")))?;
    let mut errors: Vec<_> = validator.iter_errors(document).collect();
    errors.sort_by_key(|error| error.instance_path().to_string());
    if let Some(error) = errors.first() {
        let raw_pointer = error.instance_path().to_string();
        let pointer = if raw_pointer.is_empty() {
            "<root>"
        } else {
            raw_pointer.as_str()
        };
        let message = format!("{}: {pointer}: {error}", source.display());
        return Err(if kind == "InfrastructureOutput" {
            AinfraError::output_contract(message)
        } else {
            AinfraError::input_contract(message)
        });
    }
    Ok(())
}

fn is_ipv4_network(value: &str) -> bool {
    strict_network(value, true)
}

fn is_ipv6_network(value: &str) -> bool {
    strict_network(value, false)
}

fn strict_network(value: &str, ipv4: bool) -> bool {
    let Some((address, prefix)) = value.split_once('/') else {
        return false;
    };
    let Ok(prefix) = prefix.parse::<u8>() else {
        return false;
    };
    if ipv4 {
        let Ok(address) = address.parse::<Ipv4Addr>() else {
            return false;
        };
        if prefix > 32 {
            return false;
        }
        let numeric = u32::from(address);
        let host_bits = 32_u8.saturating_sub(prefix);
        let mask = if host_bits == 32 {
            0
        } else {
            u32::MAX << host_bits
        };
        numeric & mask == numeric
    } else {
        let Ok(address) = address.parse::<Ipv6Addr>() else {
            return false;
        };
        if prefix > 128 {
            return false;
        }
        let numeric = u128::from(address);
        let host_bits = 128_u8.saturating_sub(prefix);
        let mask = if host_bits == 128 {
            0
        } else {
            u128::MAX << host_bits
        };
        numeric & mask == numeric
    }
}
